### PR TITLE
MGMT-1783 and MGMT-1784: presigned URL fixes for AWS S3

### PR DESCRIPTION
**Pull request by avishayt**

### CODE REVIEW

**Comment by romfreiman** [comment]
@avishayt where do u handle it for all our files?

**Comment by avishayt** [comment]
> @avishayt where do u handle it for all our files?

Was in a different PR, moved it here now.

File: internal/bminventory/inventory.go [hunk 1]
```diff
@@ -1486,31 +1490,27 @@ func (b *bareMetalInventory) EnableHost(ctx context.Context, params installer.En
...
 	log := logutil.FromContext(ctx, b.log)
-	var cluster common.Cluster
-	log.Infof("Download cluster files: %s for cluster %s", params.FileName, params.ClusterID)
-
-	if !funk.Contains(clusterFileNames, params.FileName) {
-		err := fmt.Errorf("invalid cluster file %s", params.FileName)
-		log.WithError(err).Errorf("failed download file: %s from cluster: %s", params.FileName, params.ClusterID)
-		return common.NewApiError(http.StatusBadRequest, err)
+	// Presigned URL only works with AWS S3 because scality is not exposed, in this case URL should be like s3.us-east-1.amazonaws.com
+	if !strings.HasSuffix(b.Config.S3EndpointURL, ".amazonaws.com") {
```
Comment (tsorya) [inline internal/bminventory/inventory.go:1496]: lets have suffix as const

**Comment by filanov** [comment]
@avishayt  Is this PR still relevant? i just tested everything with S3 and it worked for me

File: internal/bminventory/inventory.go [hunk 2]
```diff
@@ -1486,31 +1490,27 @@ func (b *bareMetalInventory) EnableHost(ctx context.Context, params installer.En
...
-				WithPayload(common.GenerateError(http.StatusInternalServerError, err))
-		}
+	if apiErr := b.checkFileForDownload(ctx, params.ClusterID.String(), params.FileName); apiErr != nil {
+		return apiErr
 	}
-	if err := b.clusterApi.DownloadFiles(&cluster); err != nil {
-		log.WithError(err).Errorf("failed to download cluster files %s", params.ClusterID)
-		return installer.NewDownloadClusterFilesConflict().
-			WithPayload(common.GenerateError(http.StatusConflict, err))
+	duration, _ := time.ParseDuration("10m")
```
Comment (tsorya) [inline internal/bminventory/inventory.go:1502]: maybe we there is a point to have it configurable?

File: internal/bminventory/inventory.go [hunk 3]
```diff
@@ -1523,31 +1523,49 @@ func (b *bareMetalInventory) DownloadClusterFiles(ctx context.Context, params in
...
+
+func (b *bareMetalInventory) checkFileForDownload(ctx context.Context, clusterID, fileName string) *common.ApiErrorResponse {
 	log := logutil.FromContext(ctx, b.log)
 	var cluster common.Cluster
-	log.Infof("Download cluster kubeconfig for cluster %s", params.ClusterID)
+	log.Infof("Checking cluster cluster file for download: %s for cluster %s", fileName, clusterID)
 
-	if err := b.db.First(&cluster, "id = ?", params.ClusterID).Error; err != nil {
-		log.WithError(err).Errorf("failed to find cluster %s", params.ClusterID)
+	if !funk.Contains(clusterFileNames, fileName) {
```
Comment (tsorya) [inline internal/bminventory/inventory.go:1543]: why clusterFileNames is not taken from swagger?

File: internal/bminventory/inventory.go [hunk 4]
```diff
@@ -366,26 +366,30 @@ func (b *bareMetalInventory) DownloadClusterISO(ctx context.Context, params inst
...
 		return errors.New("Failed to generate image: error fetching size")
 	}
+	updates["image_size_bytes"] = imgSize
+	cluster.ImageInfo.SizeBytes = &imgSize
 
-	signedURL, err := b.s3Client.GeneratePresignedDownloadURL(ctx, imgName, b.Config.ImageExpirationTime)
-	if err != nil {
-		return errors.New("Failed to generate image: error generating URL")
+	// Presigned URL only works with AWS S3 because scality is not exposed, in this case URL should be like s3.us-east-1.amazonaws.com
+	if strings.HasSuffix(b.Config.S3EndpointURL, ".amazonaws.com") {
```
Comment (filanov) [inline internal/bminventory/inventory.go:379]: empty string can be aws as well

File: internal/bminventory/inventory.go (hunk 1 above)
Comment (filanov) [inline internal/bminventory/inventory.go:1496]: maybe should be a function of the client, you have this code twice? (also empty string is also AWS)

Comment (avishayt) [reply internal/bminventory/inventory.go:1496]: Good idea

File: internal/bminventory/inventory.go (hunk 2 above)
Comment (avishayt) [reply internal/bminventory/inventory.go:1502]: Don't think so, they should download it right away

File: internal/bminventory/inventory.go (hunk 3 above)
Comment (avishayt) [reply internal/bminventory/inventory.go:1543]: I don't know, but that's how it is...

File: internal/bminventory/inventory.go [hunk 5]
```diff
@@ -1523,31 +1523,49 @@ func (b *bareMetalInventory) DownloadClusterFiles(ctx context.Context, params in
...
+
+	respBody, contentLength, err := b.s3Client.Download(ctx, fmt.Sprintf("%s/%s", params.ClusterID, kubeconfig))
+	if err != nil {
+		return installer.NewDownloadClusterKubeconfigConflict().
+			WithPayload(common.GenerateError(http.StatusConflict, errors.Wrap(err, "failed to download kubeconfig")))
+	}
+	return filemiddleware.NewResponder(installer.NewDownloadClusterKubeconfigOK().WithPayload(respBody), kubeconfig, contentLength)
+}
+
+func (b *bareMetalInventory) checkFileForDownload(ctx context.Context, clusterID, fileName string) *common.ApiErrorResponse {
```
Comment (filanov) [inline internal/bminventory/inventory.go:1538]: Usually we don't return `*common.ApiErrorResponse` in internal functions, just return an error that can be of type `*common.ApiErrorResponse` and then when you return to the user you can call to `return common.GenerateErrorResponder(err)`

File: pkg/auth/auth_handler_test.go [hunk 6]
```diff
@@ -186,6 +186,10 @@ func (f fakeInventory) DisableHost(ctx context.Context, params installer.Disable
 	panic("Implement Me!")
 }
 
+func (f fakeInventory) GetPresignedForClusterFiles(ctx context.Context, params installer.GetPresignedForClusterFilesParams) middleware.Responder {
+	panic("Implement Me!")
```
Comment (filanov) [inline pkg/auth/auth_handler_test.go:190]: its that intentional?

Comment (avishayt) [reply pkg/auth/auth_handler_test.go:190]: That's the pattern here, I just followed it...

File: internal/bminventory/inventory.go [hunk 7]
```diff
@@ -1523,31 +1523,49 @@ func (b *bareMetalInventory) DownloadClusterFiles(ctx context.Context, params in
 }
 
 func (b *bareMetalInventory) DownloadClusterKubeconfig(ctx context.Context, params installer.DownloadClusterKubeconfigParams) middleware.Responder {
+	if err := b.checkFileForDownload(ctx, params.ClusterID.String(), kubeconfig); err != nil {
+		return common.GenerateErrorResponder(err)
+	}
+
+	respBody, contentLength, err := b.s3Client.Download(ctx, fmt.Sprintf("%s/%s", params.ClusterID, kubeconfig))
+	if err != nil {
+		return installer.NewDownloadClusterKubeconfigConflict().
```
Comment (filanov) [inline internal/bminventory/inventory.go:1532]: why not just use `common.NewApiError` ?

**Pull request merged**